app/src/main/java/com/epfl/drawyourpath/userProfile/cache/UserModelCached.kt

```diff
@@ -0,0 +1,428 @@
+     *
+     * Its usage is as follow : `getUser().observe(viewLifecycleOwner) { viewText.text = it.username }`
+     * @return the [LiveData] of [UserEntity]
+     */
+    fun getUser(): LiveData<UserEntity> {
+        checkCurrentUser()
+        return user
+    }
+
+    /**
+     * get userModel
+     * @return the userModel
+     */
+    fun getUserModel(): UserModel {
+        return userModel
+    }
+
+    /**
+     * set the database used for online storing (mainly used for tests)
+     * @param database the database
+     */
+    fun setDatabase(database: Database) {
+        this.database = database
+        // if database is mock then set user to a test user
+        if (database is MockDataBase) {
+            setCurrentUser(database.userIdTest)
+        }
+    }
+
+    /**
+     * This function will return a future with a boolean to know if the username is available in the database
+     * (i.e the userName proposed is not already associated to another user profile)
+     * @param userName userName that the user want to use for his user profile
+     * @return the future that indicate if the username is available
+     */
+    fun isUsernameAvailable(userName: String): CompletableFuture<Boolean> {
+        return database.isUsernameAvailable(userName)
+    }
+
+    /**
+     * Use this function to modify the username(the username will be modify only if it is available on the database)
+     * @param username that we want to set
+     */
+    fun setUsername(username: String): CompletableFuture<Boolean> {
+        checkCurrentUser()
+        return CompletableFuture.supplyAsync {
+            database.setUsername(username).join()
+        }.thenApplyAsync {
+            if (it) {
+                cache.updateUsername(currentUserID!!, username)
+            }
+            it
+        }
+    }
+
+    /**
+     * Use this function to modify the daily distance goal of the user
+     * @param distanceGoal new daily distance goal
+     */
+    fun setDistanceGoal(distanceGoal: Double): CompletableFuture<Boolean> {
+        checkCurrentUser()
+        checkDistanceGoal(distanceGoal)
+        return CompletableFuture.supplyAsync {
+            database.setDistanceGoal(distanceGoal).join()
+        }.thenApplyAsync {
+            if (it) {
+                cache.updateDistanceGoal(currentUserID!!, distanceGoal)
+            }
+            it
+        }
+    }
+
+    /**
+     * Use this function to modify the daily activity time goal of the user
+     * @param activityTimeGoal new daily activity time goal
+     */
+    fun setActivityTimeGoal(activityTimeGoal: Double): CompletableFuture<Boolean> {
+        checkCurrentUser()
+        checkActivityTimeGoal(activityTimeGoal)
+        return CompletableFuture.supplyAsync {
+            database.setActivityTimeGoal(activityTimeGoal).join()
+        }.thenApplyAsync {
+            if (it) {
+                cache.updateTimeGoal(currentUserID!!, activityTimeGoal)

```

Comment:
Same remark here : I think it is incorrect , you should write instead : return database.setActivityTimeGoal(activityTimeGoal).thenApply{ isSet ->
if(isSet) cache.updateTimeGoal(currentUserID!!, activityTimeGoal)
isSet
}